examples/hello.rs

```diff
@@ -13,16 +13,19 @@
 // limitations under the License.
 
 use druid::shell::{runloop, WindowBuilder};
-use druid::widget::{ActionWrapper, Button, Column, DynLabel, Padding};
-use druid::{UiMain, UiState};
+use druid::widget::{ActionWrapper, Button, Column, Label, Padding};
+use druid::{LocalizedString, UiMain, UiState};
 
 fn main() {
     druid::shell::init();
 
     let mut run_loop = runloop::RunLoop::new();
     let mut builder = WindowBuilder::new();
     let mut col = Column::new();
-    let label = DynLabel::new(|data: &u32, _env| format!("value: {}", data));
+
+    let labeltext =
+        LocalizedString::new("hello-counter").with_arg("count", |_env, data: &u32| (*data).into());
+    let label = Label::new(labeltext);
```

Comment:
What I would recommend here is to design the fluent-druid binding that allows for L10nId/L10nArgs to be passed to the label, rather than eagerly resolved into a string that is passed to the label.

The way it is done now resembles the pre-fluent paradigm in Firefox and it eventually led to developers resolving strings very deep in their code and then carrying around data with resolved strings up to the UI.
This is counter-productive (if the widget doesn't get displayed, why localize it?), and it locks a lot of language selection logic deep in the imperative code without any means to "retranslate" on fly.

What I'd suggest here is sth like:

```rust
let label = Label::new();
label.set_l10n_attributes("hello-counter", fluent_args!["count" => counter);

...

fn update_counter() {
  counter += 1;
  label.set_l10n_attributes("hello-counter", fluent_args!["count" => counter);
}
```

This will bring this API much closer to how Fluent paradigm operates and in the future will allow you to bring a lot of features like runtime pseudolocales, l10n cycle indepenent of release cycle and runtime locale switching.

It also allows your layout/drawing part handle the actual localization in batches, right before the next paint, rather than synchronously string-by-string.

Reply:
I *think* what this is doing is close to what you describe? It probably isn't clear from this example how druid works, but this is just describing how to get the arguments required from the application state as needed.

Basically the label knows the fluent 'message' is it supposed to display, and for each arg required by that message, the label has a closure that can extract the value from the application state.

When the application state changes, we will automatically update the localization with the new data, and then invalidate and paint only if it has changed. Does that sound close to your suggestion?

Reply:
> When the application state changes, we will automatically update the localization with the new data, and then invalidate and paint only if it has changed.

Ahhh! That makes sense!

> Does that sound close to your suggestion?

Yes!

So, to stay closer to Fluent terminology, I'd name your `LocalizationString` a `LocalizationMessage` (for it may have more than one string in compound case), and the rest makes sense now. You pass it to `Label` and then before painting the `Label` resolve its localization. That looks good!

The one interesting piece you seem to be aiming for is something we played with as "responsive localization" where a localization is hooked to potential changes in the data state and react with retranslation.

Our main focus at the time was mobile UI l10n and we wanted to react to functions, so we had sth like:

```ftl
preferences = { SCREEN_WIDTH() ->
    [narrow] Security
   *[medium] Privacy & Security
    [wide] Privacy & Security Settings
}
```

here's a video: https://www.youtube.com/watch?v=V5K6YN7MMOA

We ended up finding this model very complicated on the Fluent bindings side and we couldn't justify the complexity at the time.
I still am intrigued by it, so if you think you can get it working, I'd say go for it!

I would still recommend having a "non-automagic" approach available. In other words, it should be possible to do:

```
label.set_l10n(l10n_id, l10n_args);
```

and have the label retranslate.

Reply:
> So, to stay closer to Fluent terminology, I'd name your `LocalizationString` a `LocalizationMessage` (for it may have more than one string in compound case), and the rest makes sense now. 

Yea, this is much closer to a fluent 'message' than anything. I'm not sure how much of fluent's naming we want to surface through our API, but at the very least this relationship should be clarified in our docs.

This also relates to the `LocalizationGroup` I was mentioning earlier; my thinking was that it might make sense to separate the compound from the non-compound case, because it could let us provide simpler APIs to our users.

> 
> The one interesting piece you seem to be aiming for is something we played with as "responsive localization" where a localization is hooked to potential changes in the data state and react with retranslation.
> 
> Our main focus at the time was mobile UI l10n and we wanted to react to functions, so we had sth like:
> 
> ...
>
> here's a video: https://www.youtube.com/watch?v=V5K6YN7MMOA

Ah, very cool. We may definitely want to explore something like this, given our cross-platform aspirations. For the time being, the use case is just things like the following, where a button increments a counter and that automatically updates a label:

![2019-08-21 13 17 07](https://user-images.githubusercontent.com/3330916/63453874-9a16fb00-c417-11e9-9cc1-ab3b96f20720.gif)

but there's obviously room for more sophistication here down the road.


> I would still recommend having a "non-automagic" approach available. In other words, it should be possible to do:
> 
> ```
> label.set_l10n(l10n_id, l10n_args);
> ```
> 
> and have the label retranslate.

We do expose an API that let's you pass a msg + args directly to the 'current bundle stack' and just get a string back; this will need some tweaking in light of attributes, but is at least a fallback option, but we're going to strongly encourage using abstractions like this `LocalizedString`; we'll see how far that gets us, I guess, and if it just isn't flexible enough we can revisit.

Reply:
> We do expose an API that let's you pass a msg + args directly to the 'current bundle stack' and just get a string back;

That's not what I suggested. This is an anti-pattern in the Fluent world.

I suggest an ability to update the id/args combo for a given UI widget.

A litmus test for when it works is:

```rust
let msg = LocalizedMessage::new("hello-counter");
let label = Label::new(msg);

fn update_label(label: &Label) {
  let l10n_id = if (User::is_logged_in()) {
     "welcome-back"
  } else {
     "welcome-new-user"
  };
  let msg = LocalizedMessage::new(l10n_id);
  label.set_l10n(msg);
}
```

This makes a clear distinction between what the engineer does (maintain the id/args status of all of the UI) from what the bindings do (they can iterate over the whole UI and retranslate at any point, or localize only right before layout etc.)

Reply:
Our label type is parameterized over its state, so for the time being I'd like to see if we can enforce the idea that you don't set args directly, you just describe how to get them. I think I understand your point, though, which is that depending on state you might want to use completely different messages. One possible solution that fits this model is that you could set the message conditionally based on some attribute in the state, but have this be declarative.

In any case thanks for the clarification, I definitely have some more thinking to do here!

Reply:
> One possible solution that fits this model is that you could set the message conditionally based on some attribute in the state, but have this be declarative.

Hmm, that may work, but it feels very complicated. As I said, our experience with trying to do stuff reactively on the state, led us to sticking to simple one-way assignments, and postponing the reactive magic for later.

Based on that experience I think it would be reasonable to provide the same basic functionality for setting/updating the l10n annotation of UI widgets independently of the state-reacting flow.

In `@fluent/dom` we provide `setAttributes` method to do so - https://github.com/projectfluent/fluent.js/blob/master/fluent-dom/src/dom_localization.js#L87 and you can see how it's used all around Firefox UI - https://searchfox.org/mozilla-central/search?q=setAttributes&path=.js

I think you can build the state-reactive model on top of the simple model that we provide, but not the other way around.

Reply:
I think I see what you're saying, and I can totally imagine having an additional API that just takes args directly; this would be helpful in circumstances where for some reason there is some state that isn't in the explicit state object, which is possible.